File: Chapter11/C#/SystemBrowser/Coordinators/CoordinatorCruise.cs
using System;
using System.ComponentModel;
using System.Windows.Forms;

namespace SystemBrowser
{
  public class CoordinatorCruise
  {
    // shortcut property
    Builder Builder
    {
      get {return Builder.Singleton;}
    }

    public void Run()
    {
      Builder.formMain.Closing += new CancelEventHandler(formMain_Closing);
      Application.Run(Builder.formMain);  // app exits when this call completes
    }

    private void formMain_Closing(object sender, System.ComponentModel.CancelEventArgs e)
    {
      // to prevent the app from closing, do this
      // e.Cancel = true;
      // return;

      // let the app shut the app down
      FireExiting();
    }

    public void HandleTopLevelEvents()
    {
      // handle the top-level events than occur during
      // the normal operation of the system...
    }

    public delegate void ExitingHandler();
    public event ExitingHandler OnExiting;
    void FireExiting()
    {
      if (OnExiting == null) return;
      OnExiting();
    }
  }
}

File: Chapter11/VB/SystemBrowser/Coordinators/CoordinatorCruise.vb
Public Class CoordinatorCruise

  ' shortcut property
  ReadOnly Property Builder() As Builder
    Get
      Return Builder.Singleton
    End Get
  End Property

  Public Sub Run()
    AddHandler Builder._formMain.Closing, AddressOf formMain_Closing
    Application.Run(Builder._formMain)  ' app exits when this call completes
  End Sub

  Private Sub formMain_Closing(ByVal sender As Object, ByVal e As System.ComponentModel.CancelEventArgs)
    'to prevent the app from closing, do this
    'e.Cancel = true;
    'return;

    ' let the app shut the app down
    FireExiting()
  End Sub

  Public Sub HandleTopLevelEvents()
    ' handle the top-level events than occur during
    ' the normal operation of the system...
  End Sub

  Public Event OnExiting()
  Sub FireExiting()
    RaiseEvent OnExiting()
  End Sub

End Class